### ISSUE TITLE
OCP HTTPS failing to load workspace.

### ISSUE THREAD

**Issue by JamesDrummond** [eclipse-che/che#8945]
### Description
Cannot get https to work on ocp. I am able to get everything to work with `ENABLE_SSL=false` but changing it to `ENABLE_SSL=false` causes workspaces to fail when loading. 

Reference issue: https://github.com/eclipse/che/issues/7798

### Reproduction Steps
The following is the cli I use to get che running on openshift ocp.

```
CHE_IMAGE_REPO="docker.io/devcomb/che-server:nightly" CHE_MULTIUSER=false \
CHE_INFRA_OPENSHIFT_USERNAME="" \
CHE_INFRA_OPENSHIFT_PASSWORD="" CHE_OPENSHIFT_PROJECT="testing123" \
OC_PUBLIC_IP=devcomb DNS_PROVIDER=com CHE_IMAGE_TAG=nightly \
CHE_PREDEFINED_STACKS_RELOAD=false CHE_DEBUG_SERVER=true \
OPENSHIFT_TOKEN=$(oc whoami -t) OPENSHIFT_ROUTING_SUFFIX="example.com" \
OPENSHIFT_ENDPOINT="https://example.com:8443" CHE_OPENSHIFT_GENERATE_SCRIPTS=false \
ENABLE_SSL=true ./ocp.sh --remove-che --deploy-che
```

I included my STAR cert file into router(documented at my blog https://medium.com/@james_devcomb/adding-wildcard-certificate-to-default-openshift-origin-router-77402464a782) and add it to service-signer.crt with associated key to service-signer.key .

`cp ~/STAR_devcomb_com.crt openshift.local.config/master/service-signer.crt && cp ~/devcomb.com.key openshift.local.config/master/service-signer.key`

I am using custom `docker.io/devcomb/che-server:nightly` and `docker.io/devcomb/ubuntu_jdk8` docker images based on eclipse che images. They both add in the STAR cert from service-signer.crt into the /etc/ssl/certs/java/cacerts . I have gone into running instances of these images in openshift to confirm that they both had these certs loaded in the java truststore /etc/ssl/certs/java/cacerts through the certs sha1 fingerprint `keytool -list -storepass changeit -keystore /etc/ssl/certs/java/cacerts` .  

Also I ran the following that produce the same sha1 signature as in the truststore file /etc/ssl/certs/java/cacerts:

`echo -n | openssl s_client -connect che-testing123.devcomb.com:443| sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > /tmp/cert-che-server.crt && openssl x509 -in /tmp/cert-che-server.crt -sha1 -noout -fingerprint `

`echo -n | openssl s_client -connect 8080-workspaceu4dbfl0cbaiyg72v.devcomb.com:443| sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > /tmp/cert-che-workspace.crt && openssl x509 -in /tmp/cert-che-workspace.crt -sha1 -noout -fingerprint ` 

**OS and version:**    
Nightly. OCP

**Diagnostics:** 
The following is the error that I get on the che server which complains about broken pipe. Not sure why it references `mb.com:8443/`?
```
2018-02-27 23:02:12,272[nio-8080-exec-7]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 228]   - Starting workspace 'che/test-wksp-af88' with id 'workspacefpwaljokt7ia0pgn' by user 'che'
--
  | 2018-02-27 23:02:21,803[mb.com:8443/...]  [ERROR] [.k.c.d.i.ExecWebSocketListener 213]  - Exec Failure: HTTP:0. Message:No response
  | java.net.SocketException: Broken pipe (Write failed)
  | at java.net.SocketOutputStream.socketWrite0(Native Method)
  | at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
  | at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
  | at sun.security.ssl.OutputRecord.writeBuffer(OutputRecord.java:431)
  | at sun.security.ssl.OutputRecord.write(OutputRecord.java:417)
  | at sun.security.ssl.SSLSocketImpl.writeRecordInternal(SSLSocketImpl.java:886)
  | at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:857)
  | at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:123)
  | at okio.Okio$1.write(Okio.java:79)
  | at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
  | at okio.RealBufferedSink.flush(RealBufferedSink.java:216)
  | at okhttp3.internal.ws.WebSocketWriter.writeControlFrameSynchronized(WebSocketWriter.java:150)
  | at okhttp3.internal.ws.WebSocketWriter.writePing(WebSocketWriter.java:79)
  | at okhttp3.internal.ws.RealWebSocket.writePingFrame(RealWebSocket.java:525)
  | at okhttp3.internal.ws.RealWebSocket$PingRunnable.run(RealWebSocket.java:513)
  | at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  | at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
  | at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
  | at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  | at java.lang.Thread.run(Thread.java:748)
Received SIGTERM
--
  | Stopping Che server running on localhost:8080
  | 2018-02-27 23:03:10,040[main]             [INFO ] [o.a.c.core.StandardServer 524]       - A valid shutdown command was received via the shutdown port. Stopping the Server instance.
  | 2018-02-27 23:03:10,042[main]             [INFO ] [o.a.c.http11.Http11NioProtocol 658]  - Pausing ProtocolHandler ["http-nio-8080"]
  | 2018-02-27 23:03:10,096[main]             [INFO ] [o.a.c.core.StandardService 477]      - Stopping service [Catalina]
  | 2018-02-27 23:03:10,273[ost-startStop-2]  [INFO ] [e.c.w.i.k.n.p.PVCSubPathHelper 197]  - PVCSubPathHelper thread pool is terminated
  | 2018-02-27 23:03:10,571[ost-startStop-2]  [INFO ] [o.e.c.a.s.server.SystemManager 92]   - Preparing system to shutdown
  | 2018-02-27 23:03:10,576[ost-startStop-2]  [INFO ] [o.e.c.a.s.s.ServiceTerminator 46]    - Shutting down 'workspace' service
  | 2018-02-27 23:03:10,584[ost-startStop-2]  [ERROR] [.s.WorkspaceServiceTermination 93]   - Couldn't get the workspace 'workspacefpwaljokt7ia0pgn' while it's running, the occurred error: 'Workspace with id 'workspacefpwaljokt7ia0pgn' doesn't exist'
```

**Comment by JamesDrummond** [comment]
@davidfestal Any thoughts on the above issue? Have you been able to get ocp working with https? If osio is working with ssl I would think there is something wrong with trusted certs or maybe a url path. Where in the code should I try to debug at?

**Comment by davidfestal** [comment]
What is the che version you're using (or branch or commit ) ?

**Comment by JamesDrummond** [comment]
Master... It's a bit old though so I will pull latest and test to be sure.

**Comment by JamesDrummond** [comment]
OK I switched to linode server which should be similar to AWS if you would like to test too. The following are my steps. I am not using any custom images in the following example.

```
export EXTERNAL_IP=$(curl -s https://4.ifcfg.me/)
openshift start master \
  --master="https://${EXTERNAL_IP}:8443" \
  --dns="https://0.0.0.0:8053" \
  --write-config="/var/lib/origin/openshift.local.config/master"
sed -i "s/router.default.svc.cluster.local/${EXTERNAL_IP}.nip.io/" \
  openshift.local.config/master/master-config.yaml
cd /var/lib/origin/
export OPENSHIFT_HOSTNAMES=kubernetes.default.svc.cluster.local,localhost,openshift.default.svc.cluster.local,127.0.0.1,172.17.0.1,172.18.0.1,172.19.0.1,172.30.0.1,192.168.122.1,192.168.42.1,$EXTERNAL_IP
oadm create-node-config \
  --dns-ip='172.30.0.1' \
  --node-dir=/var/lib/origin/openshift.local.config/node-localhost \
  --node=localhost --hostnames=$OPENSHIFT_HOSTNAMES
oc cluster up --use-existing-config --public-hostname="${EXTERNAL_IP}"
wget https://raw.githubusercontent.com/eclipse/che/master/dockerfiles/init/modules/openshift/files/scripts/ocp.sh && chmod +x ocp.sh
export OPENSHIFT_FLAVOR=ocp OPENSHIFT_TOKEN=$(oc whoami -t) OPENSHIFT_ROUTING_SUFFIX="$EXTERNAL_IP.nip.io" OPENSHIFT_ENDPOINT="https://$EXTERNAL_IP:8443" ENABLE_SSL=true
./ocp.sh --remove-che --deploy-che
```

Once I trying to create a workspace it times out but no pipe is broken as on my example above using scaleways.

Could you provide a step-by-step on AWS with ocp using HTTPS if you have been successful?

**Comment by JamesDrummond** [comment]
@davidfestal Any thoughts on the above? I have spent a lot of time trying to get this working so any help would greatly appreciated. 

Good news is I got multiuser working with http. I think earlier version was requiring https to work. I can now log into keycloak and set admin password.

**Comment by davidfestal** [comment]
The only case I tried was with a LetsEncrypt certificate added in the Che-server route using https://github.com/tnozicka/openshift-acme

Maybe @eivantsov might help ?

**Comment by eivantsov** [comment]
What's timed out? Anything helpful in Che server logs? Does ws agent start?

**Comment by JamesDrummond** [comment]
@eivantsov I had to add `CHE_INFRA_OPENSHIFT_USERNAME="" CHE_INFRA_OPENSHIFT_PASSWORD=""` to make things error out like before so now reproducible errors in linode and scaleway versions. Scaleways I did try to add my star cert file to trust store though but still having same issue of pipe broken. However what's interesting to note is that the address in the message is cut off...? For scaleways it's `mb.com:8443` instead of `devcomb.com:8443` and for linode using ipaddress it's `98.253:8443` instead of `xxx.xxx.x98.253`. Is something parsing it incorrectly or is it just shortened in the message?

``
2018-03-01 17:21:59,382[main]             [INFO ] [o.a.catalina.startup.Catalina 683]   - Server startup in 14855 ms
--
  | 2018-03-01 17:22:59,835[nio-8080-exec-3]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 461]    - Workspace 'che/wksp-f9x4' with id 'workspaceuhti62qmcs3rhzd8' created by user 'che'
  | 2018-03-01 17:23:19,193[nio-8080-exec-9]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 228]   - Starting workspace 'che/wksp-f9x4' with id 'workspaceuhti62qmcs3rhzd8' by user 'che'
  | 2018-03-01 17:23:27,129[98.253:8443/...]  [ERROR] [.k.c.d.i.ExecWebSocketListener 213]  - Exec Failure: HTTP:0. Message:No response
  | java.net.SocketException: Broken pipe (Write failed)
  | at java.net.SocketOutputStream.socketWrite0(Native Method)
  | at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
  | at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
  | at sun.security.ssl.OutputRecord.writeBuffer(OutputRecord.java:431)
  | at sun.security.ssl.OutputRecord.write(OutputRecord.java:417)
  | at sun.security.ssl.SSLSocketImpl.writeRecordInternal(SSLSocketImpl.java:886)
  | at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:857)
  | at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:123)
  | at okio.Okio$1.write(Okio.java:79)
  | at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
  | at okio.RealBufferedSink.flush(RealBufferedSink.java:216)
  | at okhttp3.internal.ws.WebSocketWriter.writeControlFrameSynchronized(WebSocketWriter.java:150)
  | at okhttp3.internal.ws.WebSocketWriter.writePing(WebSocketWriter.java:79)
  | at okhttp3.internal.ws.RealWebSocket.writePingFrame(RealWebSocket.java:525)
  | at okhttp3.internal.ws.RealWebSocket$PingRunnable.run(RealWebSocket.java:513)
  | at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  | at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
  | at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
  | at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
  | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  | at java.lang.Thread.run(Thread.java:748)
  | 2018-03-01 17:31:20,849[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 163]  - Failed to start Kubernetes runtime of workspace workspaceuhti62qmcs3rhzd8. Cause: Waiting for Kubernetes environment 'default' of the workspace'workspaceuhti62qmcs3rhzd8' reached timeout
  | 2018-03-01 17:31:57,745[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 284]   - Workspace 'che:wksp-f9x4' with id 'workspaceuhti62qmcs3rhzd8' start failed
```
